File: backend/Services/OpportunityService.cs
using System.Net.Http.Headers;
using System.Text.Json;
using System.Text.RegularExpressions;
using System.Web;
using CareerMail.Api.Data;
using CareerMail.Api.Models.DTOs;
using CareerMail.Api.Models.Entities;
using CareerMail.Api.Models.Enums;
using Microsoft.EntityFrameworkCore;

namespace CareerMail.Api.Services;

public interface IOpportunityService
{
    Task<List<OpportunityDTO>> GetOpportunitiesAsync(User user);
    Task<JobApplication> ConvertOpportunityAsync(User user, long emailId, JobApplicationRequest? customRequest);
    Task<Dictionary<string, object>> ScanGmailForOpportunitiesAsync(User user);
}

public class OpportunityService : IOpportunityService
{
    private readonly AppDbContext _context;
    private readonly IGoogleOAuthService _googleOAuthService;
    private readonly IGmailService _gmailService;
    private readonly IEmailAnalysisService _emailAnalysisService;
    private readonly HttpClient _httpClient;
    private readonly ILogger<OpportunityService> _logger;

    private static readonly string[] OpportunityKeywords = new[]
    {
        "new opportunity", "job opportunity", "career opportunity", "exciting opportunity",
        "opportunity for", "opportunity at", "opportunity with", "role opening",
        "position open", "job opening", "we are hiring", "is hiring", "hiring for",
        "thought you'd be a great fit", "thought you might be a fit", "thought of you for",
        "job match", "job alert", "recruiter reachout", "reaching out regarding",
        "wanted to reach out regarding", "found your profile", "impressed with your background",
        "open role", "new role for", "role opening at", "explore opportunities"
    };

    public OpportunityService(
        AppDbContext context,
        IGoogleOAuthService googleOAuthService,
        IGmailService gmailService,
        IEmailAnalysisService emailAnalysisService,
        HttpClient httpClient,
        ILogger<OpportunityService> logger)
    {
        _context = context;
        _googleOAuthService = googleOAuthService;
        _gmailService = gmailService;
        _emailAnalysisService = emailAnalysisService;
        _httpClient = httpClient;
        _logger = logger;
    }

    public async Task<List<OpportunityDTO>> GetOpportunitiesAsync(User user)
    {
        var allEmails = await _context.Emails
            .Include(e => e.JobApplication)
            .Where(e => e.UserId == user.Id)
            .OrderByDescending(e => e.Timestamp)
            .ToListAsync();

        var opportunities = new List<OpportunityDTO>();
        var seenSubjects = new HashSet<string>(StringComparer.OrdinalIgnoreCase);

        foreach (var email in allEmails)
        {
            var isExplicitOpportunity = email.Classification == EmailClassification.NEW_OPPORTUNITY;
            var isUnlinkedRecruiter = email.JobApplicationId == null && (
                email.Classification == EmailClassification.RECRUITER_MESSAGE ||
                email.DetectedRecruiterType == RecruiterType.HUMAN_RECRUITER
            );

            var combined = $"{email.Subject} {email.Body}".ToLowerInvariant();
            var matchesKeywords = OpportunityKeywords.Any(kw => combined.Contains(kw));

            if (isExplicitOpportunity || isUnlinkedRecruiter || (email.JobApplicationId == null && matchesKeywords))
            {
                var dedupeKey = $"{email.Subject?.Trim()}|{email.SenderEmail}";
                if (!seenSubjects.Add(dedupeKey))
                {
                    continue;
                }

                opportunities.Add(MapEmailToOpportunity(email));
            }
        }

        return opportunities;
    }

    public async Task<JobApplication> ConvertOpportunityAsync(User user, long emailId, JobApplicationRequest? customRequest)
    {
        var email = await _context.Emails
            .Include(e => e.JobApplication)
            .FirstOrDefaultAsync(e => e.Id == emailId && e.UserId == user.Id);

        if (email == null)
        {
            throw new ArgumentException($"Email not found with ID: {emailId}");
        }

        var company = !string.IsNullOrWhiteSpace(customRequest?.Company)
            ? customRequest.Company.Trim()
            : (!string.IsNullOrWhiteSpace(email.DetectedCompany) ? email.DetectedCompany : ExtractCompanyFromSender(email));

        var role = !string.IsNullOrWhiteSpace(customRequest?.Title)
            ? customRequest.Title.Trim()
            : (!string.IsNullOrWhiteSpace(email.DetectedRole) ? email.DetectedRole : "Software Engineer");

        var status = ApplicationStatus.APPLIED;
        if (!string.IsNullOrWhiteSpace(customRequest?.Status))
        {
            status = ApplicationStatusExtensions.FromString(customRequest.Status);
        }

        var app = new JobApplication
        {
            UserId = user.Id,
            Company = company,
            Title = role,
            Status = status,
            Location = customRequest?.Location ?? "Remote / Hybrid",
            EmploymentType = customRequest?.EmploymentType ?? "Full-time",
            Salary = customRequest?.Salary ?? ExtractSalary(email.Body),
            DateApplied = DateOnly.FromDateTime(DateTime.UtcNow),
            LastActivityDate = DateOnly.FromDateTime(DateTime.UtcNow),
            Source = "Gmail Opportunity Lead",
            CompanyLogo = Regex.Replace(company.ToLowerInvariant(), "[^a-z0-9]", ""),
            Priority = Priority.HIGH,
            RecruiterName = !string.IsNullOrWhiteSpace(email.DetectedRecruiterName) ? email.DetectedRecruiterName : email.Sender,
            RecruiterEmail = !string.IsNullOrWhiteSpace(email.DetectedRecruiterEmail) ? email.DetectedRecruiterEmail : email.SenderEmail,
            RecruiterType = email.DetectedRecruiterType ?? RecruiterType.HUMAN_RECRUITER,
            ContactConfidence = 85,
            CreatedAt = DateTime.UtcNow,
            UpdatedAt = DateTime.UtcNow
        };

        app.TimelineEvents.Add(new TimelineEvent
        {
            Title = "Converted Opportunity",
            Description = $"Converted from Gmail opportunity reachout: {email.Subject}",
            EventDate = DateTime.UtcNow,
            EventType = "OPPORTUNITY_CONVERTED"
        });

        _context.JobApplications.Add(app);
        await _context.SaveChangesAsync();

        email.JobApplicationId = app.Id;
        email.IsJobRelated = true;
        await _context.SaveChangesAsync();

        _logger.LogInformation("Successfully converted opportunity email ID {EmailId} into JobApplication ID {AppId} ({Company}) for user {Email}",
            emailId, app.Id, app.Company, user.Email);

        return app;
    }

    public async Task<Dictionary<string, object>> ScanGmailForOpportunitiesAsync(User user)
    {
        var account = await _context.ConnectedAccounts
            .FirstOrDefaultAsync(ca => ca.UserId == user.Id && ca.Provider == "google");

        if (account == null || string.IsNullOrWhiteSpace(account.AccessToken))
        {
            throw new InvalidOperationException("No Google account connected. Please connect Gmail in Settings first.");
        }

        var accessToken = await _googleOAuthService.GetValidAccessTokenAsync(account);
        if (string.IsNullOrWhiteSpace(accessToken))
        {
            throw new InvalidOperationException("Google authentication expired. Please reconnect Gmail in Settings.");
        }

        const string query = "newer_than:100d (\"new opportunity\" OR \"job opportunity\" OR \"career opportunity\" OR \"exciting opportunity\" OR \"opportunity with\" OR \"opportunity at\" OR \"role opening\" OR \"job opening\" OR \"we are hiring\" OR \"thought you'd be a great fit\" OR \"hiring for\")";
        var endpoint = $"https://gmail.googleapis.com/gmail/v1/users/me/messages?maxResults=60&q={HttpUtility.UrlEncode(query)}";

        var request = new HttpRequestMessage(HttpMethod.Get, endpoint);
        request.Headers.Authorization = new AuthenticationHeaderValue("Bearer", accessToken);
        var listResp = await _httpClient.SendAsync(request);

        var scanned = 0;
        var opportunitiesFound = 0;

        if (listResp.IsSuccessStatusCode)
        {
            using var listDoc = JsonDocument.Parse(await listResp.Content.ReadAsStringAsync());
            if (listDoc.RootElement.TryGetProperty("messages", out var messages) && messages.ValueKind == JsonValueKind.Array)
            {
                foreach (var m in messages.EnumerateArray())
                {
                    var msgId = m.GetProperty("id").GetString()!;
                    var threadId = m.GetProperty("threadId").GetString()!;

                    var detailUrl = $"https://gmail.googleapis.com/gmail/v1/users/me/messages/{msgId}?format=full";
                    var detailReq = new HttpRequestMessage(HttpMethod.Get, detailUrl);
                    detailReq.Headers.Authorization = new AuthenticationHeaderValue("Bearer", accessToken);
                    var detailResp = await _httpClient.SendAsync(detailReq);

                    if (detailResp.IsSuccessStatusCode)
                    {
                        using var detailDoc = JsonDocument.Parse(await detailResp.Content.ReadAsStringAsync());
                        var email = await _context.Emails.FirstOrDefaultAsync(e => e.UserId == user.Id && e.GmailMessageId == msgId);
                        if (email == null)
                        {
                            email = _gmailService.ParseGmailMessage(detailDoc.RootElement, user, msgId, threadId);
                            _context.Emails.Add(email);
                        }

                        email.Classification = EmailClassification.NEW_OPPORTUNITY;
                        email.IsJobRelated = true;
                        await _context.SaveChangesAsync();

                        await _emailAnalysisService.ProcessEmailAsync(email, user);
                        scanned++;
                        opportunitiesFound++;
                    }
                }
            }
        }

        var currentOpportunities = await GetOpportunitiesAsync(user);

        return new Dictionary<string, object>
        {
            ["success"] = true,
            ["scannedCount"] = scanned,
            ["opportunitiesFound"] = opportunitiesFound,
            ["opportunitiesCount"] = currentOpportunities.Count,
            ["message"] = $"Found {currentOpportunities.Count} extracted opportunities from your Gmail inbox.",
            ["opportunities"] = currentOpportunities
        };
    }

    private static OpportunityDTO MapEmailToOpportunity(Email email)
    {
        var company = !string.IsNullOrWhiteSpace(email.DetectedCompany)
            ? email.DetectedCompany
            : ExtractCompanyFromSender(email);

        var role = !string.IsNullOrWhiteSpace(email.DetectedRole)
            ? email.DetectedRole
            : ExtractRoleFromSubject(email.Subject);

        var recruiterName = !string.IsNullOrWhiteSpace(email.DetectedRecruiterName)
            ? email.DetectedRecruiterName
            : email.Sender;

        var recruiterEmail = !string.IsNullOrWhiteSpace(email.DetectedRecruiterEmail)
            ? email.DetectedRecruiterEmail
            : email.SenderEmail;

        var snippet = !string.IsNullOrWhiteSpace(email.Preview)
            ? email.Preview
            : (email.Body.Length > 140 ? email.Body[..140] + "..." : email.Body);

        var tags = new List<string>();
        if (email.DetectedRecruiterType == RecruiterType.HUMAN_RECRUITER)
        {
            tags.Add("Direct Recruiter Reachout");
        }
        else
        {
            tags.Add("Curated Lead");
        }
        tags.Add("Full-time");
        tags.Add("Remote / Hybrid");

        var salary = ExtractSalary(email.Body);
        var oppType = email.DetectedRecruiterType == RecruiterType.HUMAN_RECRUITER ? "Recruiter Reachout" : "Job Match Lead";

        return new OpportunityDTO(
            email.Id,
            company,
            role,
            recruiterName,
            recruiterEmail,
            email.Subject,
            snippet,
            email.Body,
            email.Timestamp,
            "Remote / Hybrid",
            salary ?? "Competitive",
            oppType,
            email.JobApplicationId.HasValue,
            email.JobApplicationId,
            tags,
            false
        );
    }

    private static string ExtractCompanyFromSender(Email email)
    {
        if (!string.IsNullOrWhiteSpace(email.Sender) && !email.Sender.Equals("Recruiter", StringComparison.OrdinalIgnoreCase) && !email.Sender.Contains("@"))
        {
            return Regex.Replace(email.Sender, @"(?i)(Careers|Recruiting|Team|Talent|Jobs)", "").Trim();
        }
        if (!string.IsNullOrWhiteSpace(email.SenderEmail) && email.SenderEmail.Contains("@"))
        {
            var domain = email.SenderEmail[(email.SenderEmail.IndexOf("@") + 1)..];
            var name = domain.Split('.')[0];
            return char.ToUpper(name[0]) + name[1..];
        }
        return "Discovered Lead";
    }

    private static string ExtractRoleFromSubject(string subject)
    {
        if (string.IsNullOrWhiteSpace(subject)) return "Software Engineer";
        var match = Regex.Match(subject, @"(?i)(?:for|as a|position:|role:)\s+([A-Za-z0-9\s/\-]+?)(?:at|with|in|\||-|$)", RegexOptions.IgnoreCase);
        if (match.Success)
        {
            var extracted = match.Groups[1].Value.Trim();
            if (extracted.Length > 3 && extracted.Length < 50)
            {
                return extracted;
            }
        }
        return "Software Engineer";
    }

    private static string? ExtractSalary(string? body)
    {
        if (string.IsNullOrWhiteSpace(body)) return null;
        var match = Regex.Match(body, @"(?i)(\$[0-9]{2,3}(?:,[0-9]{3})*(?:\s*-\s*\$[0-9]{2,3}(?:,[0-9]{3})*)?(?:\s*(?:k|k/yr|/year|per year))?)");
        return match.Success ? match.Groups[1].Value.Trim() : null;
    }
}

File: backend/app/services/opportunity_service.py
import re
from datetime import datetime, date
from sqlalchemy.orm import Session
from fastapi import HTTPException, status
from app.models.models import Email, JobApplication, TimelineEvent
from app.models.enums import EmailClassification, ApplicationStatus, RecruiterType
from app.schemas.opportunity import OpportunityDTO, OpportunityScanResult
from app.schemas.application import JobApplicationResponse
from app.services.email_analysis_service import EmailAnalysisService


class OpportunityService:
    def __init__(self, db: Session, analysis_service: EmailAnalysisService | None = None):
        self.db = db
        self.analysis_service = analysis_service or EmailAnalysisService(db)

    def get_opportunities(self, user_id: int) -> list[OpportunityDTO]:
        emails = (
            self.db.query(Email)
            .filter(
                Email.user_id == user_id,
                Email.classification == EmailClassification.NEW_OPPORTUNITY.value
            )
            .order_by(Email.timestamp.desc())
            .all()
        )

        opportunities = []
        for e in emails:
            opp = self._map_to_opportunity_dto(e)
            opportunities.append(opp)

        return opportunities

    def convert_to_application(self, user_id: int, email_id: int) -> JobApplication:
        email = (
            self.db.query(Email)
            .filter(Email.id == email_id, Email.user_id == user_id)
            .first()
        )
        if not email:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail=f"Email opportunity not found with ID: {email_id}"
            )

        if email.job_application_id:
            existing = (
                self.db.query(JobApplication)
                .filter(JobApplication.id == email.job_application_id, JobApplication.user_id == user_id)
                .first()
            )
            if existing:
                return existing

        company = email.detected_company or (email.sender if email.sender and len(email.sender.split()) <= 3 else "Unknown Company")
        title = email.detected_role or "Software Engineer"
        logo = re.sub(r"[^a-z0-9]", "", company.lower())

        app = JobApplication(
            user_id=user_id,
            company=company,
            title=title,
            location="Remote / Hybrid",
            employment_type="Full-time",
            date_applied=date.today(),
            status=ApplicationStatus.APPLIED.value,
            recruiter_name=email.detected_recruiter_name or email.sender,
            recruiter_email=email.detected_recruiter_email or email.sender_email,
            recruiter_title=email.detected_recruiter_title,
            recruiter_type=email.detected_recruiter_type or RecruiterType.HUMAN_RECRUITER.value,
            contact_confidence=email.detected_recruiter_confidence or 85,
            source="Email Opportunity Inbound",
            notes=f"Converted from inbound opportunity email: '{email.subject}'",
            company_logo=logo,
            activity_subtitle="Converted from Opportunity",
            last_activity_date=date.today(),
            created_at=datetime.utcnow(),
            updated_at=datetime.utcnow()
        )

        app.timeline_events.append(TimelineEvent(
            title=f"Converted Opportunity to Active Application",
            description=f"Inbound opportunity from {company} converted to tracked application.",
            event_date=datetime.utcnow(),
            event_type="CONVERTED_OPPORTUNITY"
        ))

        self.db.add(app)
        self.db.commit()
        self.db.refresh(app)

        email.job_application_id = app.id
        self.db.commit()

        return app

    def scan_for_opportunities(self, user_id: int) -> OpportunityScanResult:
        unprocessed_emails = (
            self.db.query(Email)
            .filter(Email.user_id == user_id)
            .all()
        )

        found_count = 0
        for e in unprocessed_emails:
            res = self.analysis_service.process_and_link_email(e, user_id)
            if res.classification == EmailClassification.NEW_OPPORTUNITY:
                found_count += 1

        self.db.commit()
        opps = self.get_opportunities(user_id)

        return OpportunityScanResult(
            success=True,
            scanned_count=len(unprocessed_emails),
            opportunities_count=len(opps),
            opportunities_found=found_count,
            message=f"Scanned {len(unprocessed_emails)} emails. Discovered {len(opps)} relevant opportunities.",
            opportunities=opps
        )

    def _map_to_opportunity_dto(self, email: Email) -> OpportunityDTO:
        company = email.detected_company or email.sender
        role = email.detected_role or "Software Engineer"
        snippet = email.preview or (email.body[:150] + "..." if len(email.body) > 150 else email.body)

        tags = ["Inbound"]
        if "remote" in email.body.lower():
            tags.append("Remote")
        if "senior" in (role or "").lower():
            tags.append("Senior")
        elif "junior" in (role or "").lower() or "graduate" in (role or "").lower():
            tags.append("Entry Level")
        else:
            tags.append("Full-time")

        return OpportunityDTO(
            id=email.id,
            company=company,
            role=role,
            recruiter_name=email.detected_recruiter_name or email.sender,
            recruiter_email=email.detected_recruiter_email or email.sender_email,
            subject=email.subject,
            snippet=snippet,
            full_body=email.body,
            body=email.body,
            received_at=email.timestamp,
            timestamp=email.timestamp,
            location="Remote / Hybrid",
            salary="Competitive",
            opportunity_type="Recruiter Reachout",
            type="Recruiter Reachout",
            is_converted=email.job_application_id is not None,
            converted=email.job_application_id is not None,
            application_id=email.job_application_id,
            tags=tags,
            is_dismissed=False,
            dismissed=False
        )
